packages/belos/tpetra/src/solvers/Belos_Tpetra_Gmres.hpp

```diff
@@ -622,35 +634,51 @@ class Gmres : public Krylov<SC, MV, OP> {
         //y.resize (restart+1);
       }
       // Compute explicit unpreconditioned residual
-      P = * (Q.getVectorNonConst (0));
-      A.apply (X, P);
-      P.update (one, B, -one);
-      r_norm = P.norm2 (); // residual norm
+      A.apply (X, R);
```

Comment:
In the future, let's talk about a fused "compute explicit residual + norm" kernel.  We've thought about this before for other solvers.